### PR TITLE
FEATURE: Use `FormDefinition` and `FieldDefinition` prototypes to seperate magic from, `attributes` and `content`

### PR DESCRIPTION
**Pull request by mficzel**
The `FormDefinition`, `FieldDefinition` and `OptionDefinition` fusion objects are now used to instantiate the helpers `form`, `field` and `option` for the context. Those helpers encapsulate all magic. So things like binding are now configured via `form.x`, `field.x` and `option.x` while `attributes.x` can set/override all attributes even the magical ones.

Changes to the fusion api by this pr: 

`Neos.Fusion.Form:FormComponent` /  `Neos.Fusion.Form:Form`
- `fieldnamePrefix` > `form.fieldnamePrefix`
- `data` > `form.data`
- `actionUri` > `form.target`
- `method` > `form.method`
- `enctype` > `form.encoding`
- `class`, `id` > removed in favor of `attributes.cass` ...

`Neos.Fusion.Form:FieldComponent`
- `name` > `field.name`
- `multiple` > `field.multiple`
- `class`, `id` > removed in favor of `attributes`

`Neos.Fusion.Form:Checkbox`
- `value` > `field.value`

`Neos.Fusion.Form:Radio`
- `value` > `field.value`

`Neos.Fusion.Form:Select.Option`
- `value` > `option.value`
- `selected` > removed in favor of generic `attributes.selected`

### CODE REVIEW

**Comment by mficzel** [comment]
I am pretty sure the api gets much more consistent but also verbose. Do you agree this is totally worth it?

I am a bit unsure wether the bool $stringify parameter on getCurrentValue() and getTargetValue() is ok or bad practice. This could be extracted into getCurrentValueStringified() and getCurentValueStringifiedArray() plus getTargetValueStringified(). Not sure which one is better and is do not like the long names.

File: Resources/Private/Fusion/Prototypes/FormComponent.fusion [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+    @propTypes {
+        form = ${PropTypes.instanceOf('\\Neos\\Fusion\\Form\\Domain\\Model\\FormHelper')}
+        attributes = ${PropTypes.dataStructure({})}
+        content = ${PropTypes.string}
+    }
+
+    #
+    # the values are used to populate the `form` context for magic
+    # The
```
Comment (skurfuerst) [inline Resources/Private/Fusion/Prototypes/FormComponent.fusion]: "the" too much :) maybe some more explanation

File: Resources/Private/Fusion/Prototypes/FormDefinition.fusion [hunk 2]
```diff
@@ -0,0 +1,20 @@
+prototype(Neos.Fusion.Form:FormDefinition) < prototype(Neos.Fusion:Component) {
```
Comment (skurfuerst) [inline Resources/Private/Fusion/Prototypes/FormDefinition.fusion]: remove inheritance here.

File: Classes/FusionObjects/FormDefinition.php [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Neos\Flow\Mvc\ActionRequest;
+use Neos\Fusion\FusionObjects\AbstractFusionObject;
+use Neos\Fusion\Form\Eel\FormHelper;
```
Comment (skurfuerst) [inline Classes/FusionObjects/FormDefinition.php]: FormHelper -> Rename to "Form"

TODO: discuss where located.

File: Classes/Eel/FormHelper.php [hunk 4]
```diff
@@ -33,102 +29,162 @@ class FormHelper implements ProtectedContextAwareInterface
...
+        if ($this->result) {
+            return $this->result->hasErrors();
+        }
+        return false;
+    }
+
+
+
     /**
      * Calculate the hidden fields for the given form content as key-value array
```
Comment (skurfuerst) [inline Classes/Eel/FormHelper.php]: TODO: add comment about why the "$content" property is still needed.

File: Resources/Private/Fusion/Prototypes/Checkbox.fusion [hunk 5]
```diff
@@ -1,25 +1,33 @@
...
-    }
 
-    checked = false
+    renderer = Neos.Fusion:Component {
+
+        #
+        # pass down props
+        #
+        attributes = ${props.attributes}
+        content = ${props.content}
```
Comment (skurfuerst) [inline Resources/Private/Fusion/Prototypes/Checkbox.fusion]: Explanation why we pass props.content down, despite not using it below (Backend Module FieldContainer e.g.)

File: Classes/Eel/AbstractFormHelper.php [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Neos\Flow\Annotations as Flow;
+use Neos\Eel\ProtectedContextAwareInterface;
+use Neos\Flow\Persistence\PersistenceManagerInterface;
+
+abstract class AbstractFormHelper implements ProtectedContextAwareInterface
```
Comment (skurfuerst) [inline Classes/Eel/AbstractFormHelper.php]: Suggestion: "Domain\AbstractForm"

**Comment by skurfuerst** [comment]
Definitions in Fusion: in "Definition" folder and namespace
Same for components.

File: Resources/Private/Fusion/Prototypes/Checkbox.fusion (hunk 5 above)
Comment (mficzel) [reply Resources/Private/Fusion/Prototypes/Checkbox.fusion]: Removed passing those down as all deriving prototyoes should at least @process the outer renderer.

**Pull request merged**